linkerd/meshtls/boring/src/server.rs

```diff
@@ -0,0 +1,177 @@
+use crate::creds::CredsRx;
+use linkerd_identity::Name;
+use linkerd_io as io;
+use linkerd_stack::{Param, Service};
+use linkerd_tls::{ClientId, LocalId, NegotiatedProtocol, ServerTls};
+use std::{future::Future, pin::Pin, sync::Arc, task::Context};
+use tracing::debug;
+
+#[derive(Clone)]
+pub struct Server {
+    name: Name,
+    rx: CredsRx,
+    alpn: Option<Arc<[Vec<u8>]>>,
+}
+
+pub type TerminateFuture<I> =
+    Pin<Box<dyn Future<Output = io::Result<(ServerTls, ServerIo<I>)>> + Send>>;
+
+#[derive(Debug)]
+pub struct ServerIo<I>(tokio_boring::SslStream<I>);
+
+// === impl Server ===
+
+impl Server {
+    pub(crate) fn new(name: Name, rx: CredsRx) -> Self {
+        Self {
+            name,
+            rx,
+            alpn: None,
+        }
+    }
+
+    pub fn with_alpn(mut self, alpn_protocols: Vec<Vec<u8>>) -> Self {
+        if !alpn_protocols.is_empty() {
+            self.alpn = Some(alpn_protocols.into());
+        }
+
```

Comment:
nit, take it or leave it: should this maybe do...something...if `alpn_protocols` is _not_ empty? maybe a warning or debug assertion? it seems like we expect this shouldn't happen, but it would be nice to actually know if it ever happens...